devito/finite_differences/utils.py

```diff
@@ -0,0 +1,139 @@
+
+    # Add non-conventional, non-centered first-order FDs
+    for d in dimensions:
+        name = 't' if d.is_Time else d.root.name
+        if function.is_Staggered:
+            # Add centered first derivatives if staggered
+            stagg = dict()
+            stagg[d] = centered
+            deriv = partial(deriv_function, deriv_order=1, dims=d,
+                            fd_order=dim_order, stagger=stagg)
+            name_fd = 'd%sc' % name
+            desciption = 'centered derivative staggered w.r.t dimension %s' % d
+
+            derivatives[name_fd] = (deriv, desciption)
+        else:
+            # Left
+            dim_order = time_fd_order if d.is_Time else space_fd_order
+            deriv = partial(deriv_function, deriv_order=1,
+                            dims=d, fd_order=dim_order, side=left)
+            name_fd = 'd%sl' % name
+            desciption = 'left first order derivative w.r.t dimension %s' % d
+            derivatives[name_fd] = (deriv, desciption)
+            # Right
+            deriv = partial(deriv_function, deriv_order=1,
+                            dims=d, fd_order=dim_order, side=right)
+            name_fd = 'd%sr' % name
+            desciption = 'right first order derivative w.r.t dimension %s' % d
+            derivatives[name_fd] = (deriv, desciption)
+
+    return derivatives
+
+
+def symbolic_weights(function, deriv_order, indices, dim):
+    return [function._coeff_symbol(indices[j], deriv_order, function, dim)
+            for j in range(0, len(indices))]
+
+
+def generate_indices(func, dim, diff, order, stagger=None, side=None):
+
+    # Check if called from first_derivative()
+    if bool(side):
+        if side == right:
+            ind = [(dim+i*diff) for i in range(-int(order/2)+1-(order % 2),
+                                               int((order+1)/2)+2-(order % 2))]
+        elif side == left:
+            ind = [(dim-i*diff) for i in range(-int(order/2)+1-(order % 2),
+                                               int((order+1)/2)+2-(order % 2))]
+        else:
+            ind = [(dim+i*diff) for i in range(-int(order/2),
+                                               int((order+1)/2)+1)]
+        x0 = None
+    else:
+        if func.is_Staggered:
+            if stagger == left:
+                off = -.5
+            elif stagger == right:
+                off = .5
+            else:
+                off = 0
+            ind = list(set([(dim + int(i+.5+off) * dim.spacing)
+                            for i in range(-order//2, order//2)]))
+            x0 = (dim + off*diff)
+            if order < 2:
+                ind = [dim + diff, dim] if stagger == right else [dim - diff, dim]
+
+        else:
+            ind = [(dim + i*dim.spacing) for i in range(-order//2, order//2 + 1)]
+            x0 = dim
+            if order < 2:
+                ind = [dim, dim + diff]
+    return ind, x0
+
+
+def form_side(dimensions, function):
```

Comment:
just noticed this: why should ``dimensions`` be different than ``function.dimensions`` ?